bitbots_localization/src/ObservationModel.cpp

```diff
@@ -4,101 +4,51 @@
-    }
-    particle_weight_goal = particle_weight_goal / goalRatings.size();
-
-  }
-
-  if (!last_measurement_field_boundary_.empty()) {
-    number_of_effective_measurements += 1;
-    fieldBoundaryRatings = map_field_boundary_->Map::provideRating(state, last_measurement_field_boundary_);
-    for (double rating : fieldBoundaryRatings) {
-      particle_weight_field_boundary += rating;
-    }
-    particle_weight_field_boundary = particle_weight_field_boundary / fieldBoundaryRatings.size();
-  }
-
-  if (!last_measurement_corners_.empty()) {
-    number_of_effective_measurements += 1;
-    cornersRating = map_corners_->Map::provideRating(state, last_measurement_corners_);
-    for (double rating : cornersRating) {
-      particle_weight_corners += rating;
+double RobotPoseObservationModel::calculate_weight_for_class(
+    const RobotState &state, 
+    const std::vector<std::pair<double, double>> &last_measurement,
+    std::shared_ptr<Map> map) const {
+  double particle_weight_for_class = 0;
+  if (!last_measurement.empty()) {
+    number_of_effective_measurements_ += 1;
+    std::vector<double> ratings = map->Map::provideRating(state, last_measurement);
+    for (double rating : ratings) {
+      particle_weight_for_class += rating;
     }
-    particle_weight_corners = particle_weight_corners / cornersRating.size();
-
-  }
-
-  if (!last_measurement_t_crossings_.empty()) {
-    number_of_effective_measurements += 1;
-    tCrossingsRating = map_t_crossings_->Map::provideRating(state, last_measurement_t_crossings_);
-    for (double rating : tCrossingsRating) {
-      particle_weight_t_crossings += rating;
-    }
-    particle_weight_t_crossings = particle_weight_t_crossings / tCrossingsRating.size();
-
+    particle_weight_for_class /= ratings.size();
   }
+  return particle_weight_for_class;
+}
 
-  if (!last_measurement_crosses_.empty()) {
-    number_of_effective_measurements += 1;
-    crossesRating = map_crosses_->Map::provideRating(state, last_measurement_crosses_);
-    for (double rating : crossesRating) {
-      particle_weight_crosses += rating;
-    }
-    particle_weight_crosses = particle_weight_crosses / crossesRating.size();
+double RobotPoseObservationModel::measure(const RobotState &state) const {
+  number_of_effective_measurements_ = 0;
 
-  }
+  double particle_weight_lines = calculate_weight_for_class(state, last_measurement_lines_, map_lines_);
+  double particle_weight_goal = calculate_weight_for_class(state, last_measurement_goal_, map_goals_);
+  double particle_weight_field_boundary = calculate_weight_for_class(state, last_measurement_field_boundary_, map_field_boundary_);
+  double particle_weight_corners = calculate_weight_for_class(state, last_measurement_corners_, map_corners_);
+  double particle_weight_t_crossings = calculate_weight_for_class(state, last_measurement_t_crossings_, map_t_crossings_);
+  double particle_weight_crosses = calculate_weight_for_class(state, last_measurement_crosses_, map_crosses_);
```

Comment:
Nice solution!

Reply:
Maybe I implement a map approach in the future, but I currently have some concerns with it.